src/scatter_buildblock/ScatterEstimationByBin.cxx

```diff
@@ -11,447 +11,1274 @@
+
+    if( this->recompute_initial_activity_image )
     {
-      error(boost::format("Error reading density_for_scatter_points image %s") %
-            this->density_image_for_scatter_points_filename);
+        info("Computing initial activity image...");
+        if (iterative_method)
+            reconstruct_iterative( -1,
+                                   this->activity_image_lowres_sptr);
+        else
+            reconstruct_analytic();
+
+        if (this->initial_activity_image_filename.size() > 0 )
+            OutputFileFormat<DiscretisedDensity < 3, float > >::default_sptr()->
+                    write_to_file(this->initial_activity_image_filename, *this->activity_image_lowres_sptr);
     }
-  this->set_density_image_for_scatter_points_sptr(this->density_image_for_scatter_points_sptr);
-}
 
+    if ( this->do_average_at_2)
+        act_image_for_averaging.reset(this->activity_image_lowres_sptr->clone());
 
-/****************** functions to set projection data **********************/
+    //
+    // Begin the estimation process...
+    //
+    for (int i_scat_iter = 0;
+         i_scat_iter < this->num_scatter_iterations;
+         i_scat_iter++)
+    {
 
-void
-ScatterEstimationByBin::
-set_template_proj_data_info_sptr(const shared_ptr<ProjDataInfo>& new_sptr)
-{
-  
-  this->proj_data_info_ptr = dynamic_cast<ProjDataInfoCylindricalNoArcCorr const *>(new_sptr->clone());
-
-  if (is_null_ptr(this->proj_data_info_ptr))
-    {     
-      error("ScatterEstimationByBin can only handle non-arccorrected data");
-    }
-      
-  // find final size of detection_points_vector
-  this->total_detectors = 
-    this->proj_data_info_ptr->get_scanner_ptr()->get_num_rings()*
-    this->proj_data_info_ptr->get_scanner_ptr()->get_num_detectors_per_ring ();
-  // reserve space to avoid reallocation, but the actual size will grow dynamically
-  this->detection_points_vector.reserve(total_detectors);
-
-  // remove any cached values as they'd be incorrect if the sizes changes
-  this->remove_cache_for_integrals_over_attenuation();
-  this->remove_cache_for_integrals_over_activity();
+        if ( this->do_average_at_2)
+        {
+            if (i_scat_iter == 2) // do average 0 and 1
+            {
+                if (is_null_ptr(act_image_for_averaging))
+                    error("Storing the first actibity estimate has failed at some point.");
+
+                //                if (this->run_debug_mode)// Write the previous estimate
+                //                {
+                //                    std::string output1= "./extras/act_image_for_averaging";
+                //                    OutputFileFormat<DiscretisedDensity<3,float> >::default_sptr()->
+                //                            write_to_file(output1, *act_image_for_averaging);
+                //                }
+                std::transform(act_image_for_averaging->begin_all(), act_image_for_averaging->end_all(),
```

Comment:
probably can do normal numerical operations

image += another_image
image /= 2

(or in 1 line)